packages/firestore/src/local/memory_persistence.ts

```diff
@@ -119,3 +154,154 @@ export class MemoryPersistence implements Persistence {
+  }
+
+  removeOrphanedDocuments(
+    txn: PersistenceTransaction,
+    upperBound: ListenSequenceNumber
+  ): PersistencePromise<number> {
+    let count = 0;
+    const cache = this.persistence.getRemoteDocumentCache();
+    const p = cache.forEachDocumentKey(txn, key => {
+      return this.isPinned(txn, key, upperBound).next(isPinned => {
+        if (isPinned) {
+          return PersistencePromise.resolve();
+        } else {
+          count++;
+          return cache.removeEntry(txn, key);
+        }
+      });
+    });
+    return p.next(() => count);
+  }
+
+  removeMutationReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  removeTarget(
+    txn: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    const updated = queryData.copy({
+      sequenceNumber: txn.currentSequenceNumber
+    });
+    return this.persistence.getQueryCache().updateQueryData(txn, updated);
+  }
+
+  addReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  removeReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  updateLimboDocument(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  private isPinned(
+    txn: PersistenceTransaction,
+    key: DocumentKey,
+    upperBound: ListenSequenceNumber
+  ): PersistencePromise<boolean> {
+    return or([
+      () => this.persistence.mutationQueuesContainKey(txn, key),
+      () => this.additionalReferences!.containsKey(txn, key),
+      () => this.persistence.getQueryCache().containsKey(txn, key),
+      () => {
+        const orphanedAt = this.orphanedSequenceNumbers.get(key);
+        return PersistencePromise.resolve(
+          orphanedAt !== undefined && orphanedAt > upperBound
+        );
+      }
+    ]);
+  }
+}
+
+/**
+ * Given an array of predicate functions that asynchronously evaluate to a
+ * boolean, implements a short-circuiting `or` between the results. Predicates
+ * will be evaluated until one of them returns `true`, then stop. The final
+ * result will be whether any of them returned `true`.
+ */
+function or(
+  predicates: Array<() => PersistencePromise<boolean>>
+): PersistencePromise<boolean> {
+  let p: PersistencePromise<boolean> = PersistencePromise.resolve(false);
+  for (let i = 0; i < predicates.length; i++) {
+    const nextPredicate = predicates[i];
```

Comment:
You don't seem to be using the index variable, so you could make this a for-each loop.

Reply:
Done.